### PR TITLE
Drive SDK

### PR DESCRIPTION
**Pull request by rohanshiva**

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -150,18 +158,24 @@ func (c *detaClient) request(i *requestInput) (*requestOutput, error) {
...
-		return nil, err
-	}
 
+	// request output
 	o := &requestOutput{
 		Status: res.StatusCode,
 		Header: res.Header,
 	}
 
+	if i.ShouldReadBody && res.StatusCode >= 200 && res.StatusCode <= 299 {
```
Comment (aavshr) [inline client.go:168]: shouldn't be called `ShouldReadBody` if you are not reading the body when `ShouldReadBody` is `true`, can be called `SkipReadBody` ?

Comment (aavshr) [reply client.go:168]: can also be called `ReturnReadCloser` probably better

File: drive.go [hunk 2]
```diff
@@ -48,12 +49,200 @@ func newDrive(projectKey, driveName, rootEndpoint string) *Drive {
...
+	if i.Name == "" {
+		return i.Name, ErrEmptyName
+	}
+
+	if i.Body == nil {
+		return i.Name, ErrEmptyData
+	}
+
+	// start upload
+	uploadId, _ := d.startUpload(i.Name)
```
Comment (aavshr) [inline drive.go:173]: we should not ignore the error

File: drive.go [hunk 3]
```diff
@@ -48,12 +49,200 @@ func newDrive(projectKey, driveName, rootEndpoint string) *Drive {
...
+
+	for {
+		chunk := make([]byte, uploadChunkSize)
+		n, err := contentStream.Read(chunk)
+		chunk = chunk[:n]
+
+		if err == io.EOF {
+			err = d.finishUpload(i.Name, uploadId)
+			if err != nil {
+				return i.Name, err
```
Comment (aavshr) [inline drive.go:185]: return the name only on successful `put` otherwise return `""`

File: drive.go [hunk 4]
```diff
@@ -48,12 +49,200 @@ func newDrive(projectKey, driveName, rootEndpoint string) *Drive {
...
+type ListOutput struct {
+	// Pagination information
+	Paging *paging `json:"paging"`
+	// list of file names
+	Names []string `json:"names"`
+}
+
+// List file names from the Drive.
+//
+// List is paginated, returns the last name fetched, and the size if further pages are left.
```
Comment (aavshr) [inline drive.go:210]: only `last` determines if further pages left, `size` is always returned

File: drive_test.go [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+)
+
+const (
+	readChunkSize = 1024 * 1024 * 10
+)
+
+func SetupDrive() *Drive {
+	projectKey := os.Getenv("DETA_SDK_TEST_PROJECT_KEY")
+	driveName := os.Getenv("DETA_SDK_TEST_DRIVE_NAME")
+	rootEndpoint := os.Getenv("DETA_SDK_TEST_ENDPOINT")
```
Comment (aavshr) [inline drive_test.go:21]: should be `DETA_SDK_TEST_DRIVE_ENDPOINT`

File: drive_test.go [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+		}
+	}
+
+}
+
+func TestPutDrive(t *testing.T) {
+	drive := SetupDrive()
+	defer TearDownDrive(drive, t)
+
+	testCases := []struct {
```
Comment (aavshr) [inline drive_test.go:43]: can you please also put test cases with
1. empty content type
2. other io.Readers besides `strings.NewReader`

**Review by aavshr — CHANGES_REQUESTED** [review]
thanks @rohanshiva , left come comments. Also additionally, 
1. If we are returning an `error`, we generally should return an empty value for the other return value
2. `reflect.DeepEqual` can have some pitfalls sometimes, please make sure to check if it's being used correctly.

**Comment by rohanshiva** [comment]
Thanks @aavshr, I will check those.

File: drive_test.go [hunk 7]
```diff
@@ -0,0 +1,361 @@
+package deta
+
+import (
```
Comment (aavshr) [inline drive_test.go:3]: missing import "os"

File: drive_test.go [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+	"bytes"
+	"crypto/rand"
```
Comment (aavshr) [inline drive_test.go:5]: `math/rand` is better, we don't need crypto, crypto rand read is slower

**Review by aavshr — APPROVED** [review]
thanks @rohanshiva , approved, one last small thing with using `math/rand` instead of `crypto/rand`
and the test failing. 

Can you please check them out?

**Comment by rohanshiva** [comment]
thanks @aavshr, changed to `math/rand` and the tests run successfully.

**Pull request merged**